### DISCUSSION TITLE
CURL Example

### DISCUSSION THREAD

**Discussion by CaffeineLab** [ercot/api-specs#66] [Show and tell]
You will need your login username and password as well as your subscription key for your ERCOT API Account.

Get an Authentication Token
POST (not get) the ENTIRE request below for a token, make sure to update the following URL with the username and password.  Yes, scope and client_id are required, NO they do not change.  

`curl --location --request POST "https://ercotb2c.b2clogin.com/ercotb2c.onmicrosoft.com/B2C_1_PUBAPI-ROPC-FLOW/oauth2/v2.0/token?username=************&password=************&grant_type=password&scope=openid+fec253ea-0d06-4272-a5e6-b478baeecd70+offline_access&client_id=fec253ea-0d06-4272-a5e6-b478baeecd70&response_type=id_token
`

This will return a JSON document 

```
{
    "access_token": "",
    "token_type": "Bearer",
    "expires_in": "3600",
    "refresh_token": "",
    "id_token": "THIS IS THE ONE YOU WANT - and it is rather long...  1080 chars even..."
}
```

Now you are ready to make calls to the report.  

`curl --location --request GET "https://api.ercot.com/api/public-reports/np6-346-cd/act_sys_load_by_fzn?operatingDayFrom=2023-01-01&operatingDayTo=2023-12-31"  --header "Host: api.ercot.com" --header "Accept: application/json" --header "Ocp-Apim-Subscription-Key: **YOUR*SUBSCRIPTION*KEY*FROM*ACCOUNT**" --header "Authorization: Bearer ***id_token_from_above***"
`

Important: Most applications will automatically generate the Host header for you, without it the server responds with a 503.
Important: If you modify the client_id in the token request you will get a 400 error stating that the client id is not registered in the tenant.  
Important: The report URL is all lowercase... if you use an uppercase emil id, eg NP6-346-CD in the URL, you will get a 404.

**Comment by DMJ-CV-91913** [comment]
Thanks, I retrieve the access token then run another curl request, and at that point, I do not retrieve a json response.

I was able to get a response for archives, document links, etc.    But not with public reports.  I verified registration this morning, and still no response.  (ie)

https://api.ercot.com/api/public-reports/NP4-732-CD/wpp_hrly_avrg_actl_fcast
https://api.ercot.com/api/public-reports/NP4-732-CD/wpp_hrly_avrg_actl_fcast?deliveryDateFrom=2024-01-01&deliveryDateTo=2024-10-31

I pass headers for subscription key, and access token in a CURL request and receive no response.

I responded with my code in another link already.  https://github.com/ercot/api-specs/discussions/34#discussioncomment-11438993

**Reply by CaffeineLab** [reply]
Try lowercase

https://api.ercot.com/api/public-reports/np4-732-cd/wpp_hrly_avrg_actl_fcast

**Reply by DMJ-CV-91913** [reply]
yes that produced a response.  Thank you.  I will let you know if any other issues.  

One possible suggestion on the developers forum would be to display the emilID report name as lowercase, because the report name is currently in uppercase, but upon review the url path is lowercase.  That was my error, but does lend to some possible confusion.

https://developer.ercot.com/applications/pubapi/user-guide/openapi-documentation/

/np4-732-cd/wpp_hrly_avrg_actl_fcast

NP4-732-CD

I tried to attach a screen shot  ".png " but it wouldnt let me add the file.

**Reply by mackermann-ercot** [reply]
DMJ-CV-91913, 

I will look into either changing the documentation or allowing the endpoint to ignore case for the EMIL Product identifier.  

Regards,
mackermann-ercot

**Comment by loamylo** [comment]
Hello,

I'm trying to request np3-965-er/60_sced_gen_res_data through CURL but nothing I've tried is working. I know it's the structure of the parameters but I don't know how to fix it.  I've had success with the np4-190-cd/dam_stlmnt_pnt_prices.  Any ideas what I am doing wrong?

THIS WORKS
 curl --location --request GET "https://api.ercot.com/api/public-reports/np4-190-cd/dam_stlmnt_pnt_prices?deliveryDateFrom=2026-05-18&download=csv" --header "Ocp-Apim-Subscription-Key: XXXXXXXXXXXX" --header "Authorization: Bearer XXXXXXXXX" -H "accept: application/json" -k -o "C:\damPrices.zip"

THIS DOES NOT WORK
curl --location --request GET "https://api.ercot.com/api/public-reports/np3-965-er/60_sced_gen_res_data?resourceName=FRYE_SLR_UNIT1&SCEDTimestampFrom=2026-05-01T00%3A00%3A00&SCEDTimestampTo=2026-05-02T00%3A13%3A59&size=30000" --header "Ocp-Apim-Subscription-Key: XXXXXXXXXXXXX" --header "Authorization: Bearer XXXXXXXXXXXXXXXXXXXXXXXXX" -H "accept: application/json" -k -o "C:\genResource.zip"

THIS ALSO DOES NOT WORK
curl -location --request GET "https://api.ercot.com/api/public-reports/np3-965-er/60_sced_gen_res_data?SCEDTimestampFrom=2026-05-01T00:00:00&SCEDTimestampTo=2026-05-02T00:00:00&size=30000" --header "Ocp-Apim-Subscription-Key: 917510bc9dea498a81c7d4b695d7172b" --header "Authorization: Bearer XXXXX" -H "accept: application/json" -k -o "C:\genResource.zip"
